packages/api/lambdas/sf-event-sqs-to-db-records.js

```diff
@@ -1,22 +1,95 @@
 
-const saveExecutionToDb = async (cumulusMessage) => {
+const isPostRDSDeploymentExecution = (cumulusMessage) => {
+  const minimumSupportedRDSVersion = process.env.RDS_DEPLOYMENT_CUMULUS_VERSION;
+  if (!minimumSupportedRDSVersion) {
+    throw new Error('RDS_DEPLOYMENT_CUMULUS_VERSION environment variable must be set');
+  }
+  const cumulusVersion = get(cumulusMessage, 'cumulus_meta.cumulus_version', '0.0.0');
+  return semver.gte(cumulusVersion, minimumSupportedRDSVersion);
+};
+
+/**
+ * Get the initial execution message for a chain of workflows.
+ *
+ * @param {Object} cumulusMessage - A Cumulus workflow message
+ * @returns {Object}
+ *   Cumulus workflow message for the parent execution initiating
+ *   this chain of executions.
+ */
+const getInitialExecutionMessage = async (cumulusMessage) => {
+  const parentArn = getMessageExecutionParentArn(cumulusMessage);
+  if (!parentArn) {
+    return cumulusMessage;
+  }
+  const executionArn = getMessageExecutionArn(cumulusMessage);
+  const response = await StepFunctions.describeExecution({
+    executionArn,
+  });
+  return getInitialExecutionMessage(JSON.parse(response.input));
+};
+
+const shouldWriteToRDS = async (cumulusMessage) => {
+  const initialMessage = await getInitialExecutionMessage(cumulusMessage);
+  return isPostRDSDeploymentExecution(initialMessage);
+};
+
+const saveExecutionToDynamoDb = async (cumulusMessage, executionModel) =>
+  executionModel.storeExecutionFromCumulusMessage(cumulusMessage);
+
+const deleteExecutionFromRDS = async (isRDSWriteEnabled, executionArn, knex) => {
+  if (!isRDSWriteEnabled) return true;
+  return knex('executions')
+    .where({
+      arn: executionArn,
+    })
+    .delete();
+};
+
+// Just a stub for write functionality
+const saveExecutionToRDS = async (isRDSWriteEnabled, cumulusMessage, knex) => {
+  if (!isRDSWriteEnabled) return true;
+  const executionArn = getMessageExecutionArn(cumulusMessage);
+  return knex('executions').insert({
+    arn: executionArn,
+  });
+};
+
+const saveExecutions = async (cumulusMessage, isRDSWriteEnabled, knex) => {
   const executionModel = new Execution();
+  const executionArn = getMessageExecutionArn(cumulusMessage);
   try {
-    await executionModel.storeExecutionFromCumulusMessage(cumulusMessage);
+    await pAll([
+      () => saveExecutionToDynamoDb(cumulusMessage, executionModel),
+      () => saveExecutionToRDS(isRDSWriteEnabled, cumulusMessage, knex),
+    ], {
+      // let all promises settle before throwing error so that we know
+      // all writes have either failed/succeeded
+      stopOnError: false,
+    });
   } catch (error) {
-    const executionArn = getMessageExecutionArn(cumulusMessage);
-    log.fatal(`Failed to create/update database record for execution ${executionArn}: ${error.message}`);
-    throw error;
+    log.error(`Failed to write execution records for ${executionArn}`, error);
```

Comment:
There may be fragility here if the deletion fails for \<AWS reasons\>, particularly given how buried this lambda is in the message handling.  

A couple of other thoughts: 

* re: DynamoDb consistency - a situation where saveExecutionToRDS fails, saveExecutionToDynamoDb succeeds and the deletions run quickly enough that executionModel.delete thinks it deleted things - I've not looked into that scenario with Dynamo - is it possible that read consistency results in a false positive on delete for a table under load?  

* In a production/demand situation where a large number of failures with result in read/write retries might have unintended consequences/performance concerns.     I need to think about this more, but wanted to get the thought down for consideration.

Reply:
Longer term, really doing an async separate write here probably doesn't make sense anyway - should we consider something along the lines of transactional RDS write/fail, then run the dynamo writes if that succeeds to avoid the issue entirely?

Doubly, shouldn't some of the RDS read/write be baked into the model?   I realize `saveExecutionToRDS` is just a stub for now, but what is that going to look like moving forward?

Reply:
Re: transactional RDS/write fail - I'm not opposed at all, but what's the benefit of that approach. My understanding is that we want to delete **both the RDS and Dynamo records** if **either write fails**. Now that I start talking, I suppose the better approach would be starting a transaction for the RDS write, and having that transaction rollback if the Dynamo write fails?

As for using "models", I don't think we ever established how we wanted to approach that. I remember Huffnagle having some opposition to re-creating "models" for managing the interaction with RDS, though I don't know why. This feels like a good architecture meeting topic.

Reply:
This has all been extensively refactored to use transactional writes. Based on how the transactions work in Knex, the Dynamo write is only even attempted if the RDS write succeeds. And the RDS write is only committed if the Dynamo write succeeds. So I think that covers the write consistency?

Knex docs on transactions: http://knexjs.org/#Transactions